lib/ansible/modules/network/onyx/onyx_traffic_class.py

```diff
@@ -0,0 +1,329 @@
+    IF_PO_REGEX = re.compile(r"^Po(\d+)$")
+    MLAG_NAME_REGEX = re.compile(r"^Mpo(\d+)$")
+
+    IF_TYPE_ETH = "ethernet"
+    PORT_CHANNEL = "port-channel"
+    MLAG_PORT_CHANNEL = "mlag-port-channel"
+
+    IF_TYPE_MAP = {
+        IF_TYPE_ETH: IF_ETH_REGEX,
+        PORT_CHANNEL: IF_PO_REGEX,
+        MLAG_PORT_CHANNEL: MLAG_NAME_REGEX
+    }
+
+    def init_module(self):
+        """ initialize module
+        """
+        congestion_control_spec = dict(control=dict(choices=['red', 'ecn', 'both'], required=True),
+                                       threshold_mode=dict(choices=['absolute', 'relative'], required=True),
+                                       min_threshold=dict(type=int, required=True),
+                                       max_threshold=dict(type=int, required=True))
+
+        dcb_spec = dict(mode=dict(choices=['strict', 'wrr'], required=True),
+                        weight=dict(type=int))
+
+        element_spec = dict(
+            interfaces=dict(type='list', required=True),
+            tc=dict(type=int, required=True),
+            congestion_control=dict(type='dict', options=congestion_control_spec),
+            dcb=dict(type='dict', options=dcb_spec),
+            state=dict(choices=['enabled', 'disabled'], default='enabled'))
+
+        argument_spec = dict()
+        argument_spec.update(element_spec)
+        self._module = AnsibleModule(
+            argument_spec=argument_spec,
+            supports_check_mode=True)
+
+    def get_required_config(self):
+        module_params = self._module.params
+        self._required_config = dict(module_params)
+        self.validate_param_values(self._required_config)
+
+    def validate_tc(self, value):
+        if value and not 0 <= int(value) <= 7:
+            self._module.fail_json(msg='tc value must be between 0 and 7')
+
+    def validate_param_values(self, obj, param=None):
+        dcb = obj.get("dcb")
+        if dcb is not None:
+            dcb_mode = dcb.get("mode")
+            weight = dcb.get("weight")
+            if dcb_mode == "wrr" and weight is None:
+                self._module.fail_json(msg='User should send weight attribute when dcb mode is wrr')
+        super(OnyxTrafficClassModule, self).validate_param_values(obj, param)
+
+    def _get_interface_type(self, if_name):
+        if_type = None
+        if_id = None
+        for interface_type, interface_regex in iteritems(self.IF_TYPE_MAP):
+            match = interface_regex.match(if_name)
+            if match:
+                if_type = interface_type
+                if_id = match.group(1)
+                break
+        return if_type, if_id
+
+    def _set_interface_congestion_control_config(self, interface_congestion_control_config,
+                                                 interface, if_type, if_id):
+        tc = self._required_config.get("tc")
+        interface_dcb_ets = self._show_interface_dcb_ets(if_type, if_id)[0].get(interface)
```

Comment:
what `_show_interface_dcb_ets(if_type, if_id)` fails:
better to handle the case if the return value is None

Reply:
Done